x-pack/filebeat/input/etw/input.go

```diff
@@ -0,0 +1,215 @@
+	}
+}
+
+func configure(cfg *conf.C) (stateless.Input, error) {
+	conf := defaultConfig()
+	if err := cfg.Unpack(&conf); err != nil {
+		return nil, err
+	}
+
+	return &etwInput{
+		config: conf,
+	}, nil
+}
+
+func (e *etwInput) Name() string { return inputName }
+
+func (e *etwInput) Test(_ input.TestContext) error {
+	// ToDo
+	return nil
+}
+
+// Run starts the ETW session and processes incoming events.
+func (e *etwInput) Run(ctx input.Context, publisher stateless.Publisher) error {
+	var err error
+	// Initialize a new ETW session with the provided configuration.
+	e.etwSession, err = etw.NewSession(convertConfig(e.config))
+	if err != nil {
+		return fmt.Errorf("error when initializing '%s' session: %w", e.etwSession.Name, err)
+	}
+
+	// Set up logger with session information.
+	e.log = ctx.Logger.With("session", e.etwSession.Name)
+	e.log.Info("Starting " + inputName + " input")
+
+	var wg sync.WaitGroup
+	var once sync.Once
+
+	// Handle realtime session creation or attachment.
+	if e.etwSession.Realtime {
+		if !e.etwSession.NewSession {
+			// Attach to an existing session.
+			err = e.etwSession.GetHandler()
+			if err != nil {
+				return fmt.Errorf("unable to retrieve handler: %w", err)
+			}
+			e.log.Debug("attached to existing session")
+		} else {
+			// Create a new realtime session.
+			err = e.etwSession.CreateRealtimeSession()
+			if err != nil {
+				return fmt.Errorf("realtime session could not be created: %w", err)
+			}
+			e.log.Debug("created session")
+		}
+	}
+	// Defer the cleanup and closing of resources.
+	defer func() {
+		wg.Wait() // Ensure all goroutines have finished before closing.
+		once.Do(e.Close)
+		e.log.Info(inputName + " input stopped")
+	}()
+
+	// eventReceivedCallback processes each ETW event.
+	eventReceivedCallback := func(er *etw.EventRecord) uintptr {
+		if er == nil {
+			e.log.Error("received null event record")
+			return 1
+		}
+
+		e.log.Debugf("received event %d with length %d", er.EventHeader.EventDescriptor.Id, er.UserDataLength)
+
+		var event map[string]interface{}
+
+		if data, err := etw.GetEventProperties(er); err == nil {
+			event = map[string]interface{}{
+				"Header":          er.EventHeader,
+				"EventProperties": data,
+				"Metadata":        fillEventMetadata(er, e.etwSession, e.config),
+			}
+		} else {
+			e.log.Errorf("failed to read event properties: %w", err)
+			return 1
+		}
+
+		evt := beat.Event{
+			Timestamp: time.Now(),
+			Fields: mapstr.M{
+				"metadata": event["Metadata"],
+				"header":   event["Header"],
+				"winlog":   event["EventProperties"],
+			},
+		}
+		publisher.Publish(evt)
+
+		return 0
+	}
+
+	// Set the callback function for the ETW session.
+	e.etwSession.Callback = syscall.NewCallback(eventReceivedCallback)
+
+	// Start a goroutine to consume ETW events.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		e.log.Debug("starting to listen ETW events")
+		if err = e.etwSession.StartConsumer(); err != nil {
```

Comment:
Are there any errors that can occur while reading that should be retried? In my experience most inputs require some type of retry logic or watchdog that restarts the reading upon error. Otherwise some basic error can stop the input and the user will not receive more logs until the restart Filebeat.

Reply:
I had this on mind during the development. Honestly I didn't identify a clear situation where a retry mechanism would fit so far.

At `StartConsumer` there are some previous steps before starting to consume events, where every raised error is not recoverable by itself (permissions, misconfiguration, etc.). Once started the reading of events, that thread gets blocked and events are forwarded to the callback, where a failure in the parse of an event doesn't throw the input down but it only affect the event that is being processed. Also I couldn't find any reference to retryable situations in docs:
- https://learn.microsoft.com/en-us/windows/win32/etw/consuming-events
- https://learn.microsoft.com/en-us/windows/win32/api/evntrace/nc-evntrace-pevent_record_callback
- https://learn.microsoft.com/en-us/windows/win32/api/evntrace/nf-evntrace-processtrace

Did you have any particular situation in mind?